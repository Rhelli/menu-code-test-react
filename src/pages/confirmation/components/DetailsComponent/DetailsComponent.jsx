import React from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import dateFormat from 'dateformat';
import { extractOrderDetails, extractSubTotal, extractGrandTotal } from '../../../../utils/menuUtils';
import styles from './DetailsComponent.module.scss';

const DetailsComponent = ({ orderStore }) => {
  const { partySize, time, date } = orderStore.booking;
  const { orders } = orderStore;
  const names = Object.keys(orders);

  return (
    <div className={styles.detailsContainer}>
      <div className={styles.bookingDetails}>
        <div>
          <h5>Party Size</h5>
          <p>
            For
            {' '}
            {partySize}
          </p>
        </div>
        <div>
          <h5>Time</h5>
          <p>
            {time}
          </p>
        </div>
        <div>
          <h5>Date</h5>
          <p>
            {dateFormat(date, 'dddd, dS mmmm')}
          </p>
        </div>
      </div>
      <h3>Your Orders</h3>
      <div className={styles.orderList}>
        <div>
          {
            names.map((name) => (
              <div key={uuidv4()} className={styles.guestOrder}>
                <h5>{name}</h5>
                {
                  extractOrderDetails(orders, name).map((item) => (
                    <div key={uuidv4()} className={styles.orderItem}>
                      <p>{item[0]}</p>
                      <p className={styles.price}>{item[1]}</p>
                    </div>
                  ))
                }
                <div className={styles.individualTotal}>
                  {
                    name !== 'Sharing' ? (
                      <p>
                        {name}
                        &apos;s
                        {' '}
                        total:
                      </p>
                    ) : (
                      <p>
                        {name}
                        {' '}
                        total:
                      </p>
                    )
                  }
                  <p className={styles.total}>
                    {extractSubTotal(orders, name)}
                  </p>
                </div>
              </div>
            ))
          }
        </div>
        <div className={styles.grandTotalContainer}>
          <p>Grand Total:</p>
          <p className={styles.grandPrice}>
            {extractGrandTotal(orders)}
          </p>
        </div>
      </div>
    </div>
  );
};

DetailsComponent.propTypes = {
  orderStore: PropTypes.shape({
    orders: PropTypes.object,
    names: PropTypes.arrayOf(PropTypes.string),
    booking: PropTypes.shape({
      partySize: PropTypes.number.isRequired,
      time: PropTypes.string.isRequired,
      date: PropTypes.string.isRequired
    })
  }).isRequired
};

export default DetailsComponent;
