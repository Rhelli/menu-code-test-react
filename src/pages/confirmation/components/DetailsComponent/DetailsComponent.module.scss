.detailsContainer {
  display: flex;
  width: 100%;
  flex-flow: column nowrap;

  .bookingDetails {
    display: flex;
    flex-flow: row nowrap;
    width: 100%;
    margin-bottom: 10px;

    div {
      flex-grow: 1;
      height: auto;
      padding: 10px 0;
  
      h5 {
        font-size: 16px;
      }
  
      p {
        padding: 10px 0;
        border-bottom: 1px solid #d8d9db;
      }
    }
  }

  .orderList {
    
    .guestOrder {
      display: flex;
      flex-flow: column nowrap;
      width: 100%;
      padding: 10px 0;

      h5 {
        margin-bottom: 5px;
        font-size: 16px;
      }

      .orderItem {
        display: flex;
        flex-flow: row nowrap;
        padding: 4px 0;

        p {
          font-size: 14px;
        }

        .price {
          margin-left: auto;
        }
      }
    }

    .individualTotal {
      display: flex;
      flex-flow: row nowrap;
      align-items: center;
      width: 100%;
      padding: 6px 0;
      border-top: 1px solid #da3743;
      border-bottom: 1px solid #da3743;
      margin-top: 10px;

      .total {
        margin-left: auto;
      }
    }

    .grandTotalContainer {
      display: flex;
      flex-flow: row nowrap;
      width: 100%;
      padding: 20px 0;
      border-bottom: 1px solid #d8d9db;

      p {
        font-size: 20px;
      }

      .grandPrice {
        margin-left: auto;
      }
    }
  }

  .confirmButton {
    display: flex;
    width: 100%;
    align-items: center;
    justify-content: center;
    padding: 30px 0;

    button {
      background-color: #da3743;
      border: none;
      outline: none;
      color: #fff;
      width: 45%;
      height: 45px;
      border-radius: 3px;
      transition-duration: 0.1s;
      font-size: 16px;

      &:hover {
        background-color: #A42932;
        transition-duration: 0.15s;
      }
    }
  }
}