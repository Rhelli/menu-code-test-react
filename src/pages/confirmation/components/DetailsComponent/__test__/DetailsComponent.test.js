import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import DetailsComponent from '../DetailsComponent';

Enzyme.configure({ adapter: new Adapter() });

const orderStore = {
  booking: {
    partySize: 2,
    time: '12:00',
    date: '2021-06-10T13:16:00.983Z'
  },
  orders: {
    Sam: {
      color: {
        backgroundColor: '#d3445d'
      },
      starters: {
        food: 'Lambas Bread',
        price: '5.00'
      },
      mains: {
        food: 'Taters',
        price: '11.50'
      },
      desserts: {}
    },
    Frodo: {
      color: {
        backgroundColor: '#404040'
      },
      starters: {
        food: 'Lambas Bread',
        price: '5.00'
      },
      mains: {
        food: 'Second Breakfast',
        price: '13'
      },
      desserts: {}
    }
  }
}

const detailsComponent = (<DetailsComponent orderStore={orderStore} />);

describe('The appearance and functionality of the DetailsComponent', () => {
  const container = mount(<DetailsComponent orderStore={orderStore} />);

  it('Matches the previous snapshot taken', () => {
    const tree = renderer.create(detailsComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders correctly', () => {
    mount(<DetailsComponent orderStore={orderStore} />);
  });

  it('Displays the booking details correctly with a formatted date', () => {
    expect(container.find('.bookingDetails').childAt(0).childAt(1).text()).toBe('For 2');
    expect(container.find('.bookingDetails').childAt(1).childAt(1).text()).toBe('12:00');
    expect(container.find('.bookingDetails').childAt(2).childAt(1).text()).toBe('Thursday, 10th June');
  });
});
