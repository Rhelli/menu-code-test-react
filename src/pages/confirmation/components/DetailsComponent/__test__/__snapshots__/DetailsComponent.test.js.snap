// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The appearance and functionality of the DetailsComponent Matches the previous snapshot taken 1`] = `
<div
  className="detailsContainer"
>
  <div
    className="bookingDetails"
  >
    <div>
      <h5>
        Party Size
      </h5>
      <p>
        For
         
        2
      </p>
    </div>
    <div>
      <h5>
        Time
      </h5>
      <p>
        12:00
      </p>
    </div>
    <div>
      <h5>
        Date
      </h5>
      <p>
        Thursday, 10th June
      </p>
    </div>
  </div>
  <h3>
    Your Orders
  </h3>
  <div
    className="orderList"
  >
    <div>
      <div
        className="guestOrder"
      >
        <h5>
          Sam
        </h5>
        <div
          className="orderItem"
        >
          <p>
            Lambas Bread
          </p>
          <p
            className="price"
          >
            £5.00
          </p>
        </div>
        <div
          className="orderItem"
        >
          <p>
            Taters
          </p>
          <p
            className="price"
          >
            £11.00
          </p>
        </div>
        <div
          className="individualTotal"
        >
          <p>
            Sam
            's
             
            total:
          </p>
          <p
            className="total"
          >
            £16.00
          </p>
        </div>
      </div>
      <div
        className="guestOrder"
      >
        <h5>
          Frodo
        </h5>
        <div
          className="orderItem"
        >
          <p>
            Lambas Bread
          </p>
          <p
            className="price"
          >
            £5.00
          </p>
        </div>
        <div
          className="orderItem"
        >
          <p>
            Second Breakfast
          </p>
          <p
            className="price"
          >
            £13.00
          </p>
        </div>
        <div
          className="individualTotal"
        >
          <p>
            Frodo
            's
             
            total:
          </p>
          <p
            className="total"
          >
            £18.00
          </p>
        </div>
      </div>
    </div>
    <div
      className="grandTotalContainer"
    >
      <p>
        Grand Total:
      </p>
      <p
        className="grandPrice"
      >
        £34.00
      </p>
    </div>
  </div>
</div>
`;
